services/catalog/rs_server_catalog/user_catalog.py

```diff
@@ -341,48 +343,165 @@ async def manage_put_post_request(self, request: Request) -> Request | JSONRespo
+        i = 0
+        # Delete all collections that the user does not have access to.
+        while i < len(content["links"]):
+            link = content["links"][i]
+
+            if link["rel"] == "child":  # Only check child links.
+                authorized = 0
+                for role in auth_roles:  # Check in the authorisations list if the link is allowed to the user.
+                    if match := re.match(catalog_read_right_pattern, role):
+                        groups = match.groupdict()
+                        if (
+                            groups["owner_id"] in link["href"]
+                            and (groups["collection_id"] == "*" or groups["collection_id"] in link["href"])
+                            and groups["right_type"] == "read"
+                        ):
+                            authorized = 1
+                            break
+                if not authorized:
+                    del content["links"][i]
+                else:
+                    i += 1
+            else:
+                i += 1
+        return content
+
+    def manage_all_collections(self, collections: dict, auth_roles: list, user_login: str) -> list:
+        """Return the list of all collections accessible by the user calling it.
+
+        Args:
+            collections (dict): List of all collections.
+            auth_roles (list): List of roles of the api-key.
+            user_login (str): The api-key owner.
+
+        Returns:
+            dict: The list of all collections accessible by the user.
+        """
+        accessible_collections = []
+        catalog_read_right_pattern = (
+            r"rs_catalog_(?P<owner_id>.*(?=:)):(?P<collection_id>.+)_(?P<right_type>read|write|download)(?=$)"
+        )
+        for role in auth_roles:
+            if match := re.match(catalog_read_right_pattern, role):
+                groups = match.groupdict()
+                if groups["right_type"] == "read":  # The user calling the endpoint as read access to:
+                    if groups["collection_id"] == "*":  # The user calling the endpoint has access to the entire catalog
+                        new_accessible_collections = filter_collections(collections, groups["owner_id"])
+                        accessible_collections += new_accessible_collections
+                    else:
+                        # The user calling the endpoint has access
+                        # to a specific collection from the catalog group["owner_id"]
+                        new_accessible_collection = filter_collections(
+                            collections,
+                            f"{groups['owner_id']}_{groups['collection_id']}",
+                        )
+                        accessible_collections += new_accessible_collection
+        accessible_collections += filter_collections(collections, user_login)
+        return accessible_collections
```

Comment:
Same here, maybe a more python approach:

`accessible_collections = []

# Filter roles for read access
read_roles = [role for role in auth_roles if re.match(catalog_read_right_pattern, role)]

for role in read_roles:
    groups = re.match(catalog_read_right_pattern, role).groupdict()
    if groups["right_type"] == "read":
        owner_id = groups["owner_id"]
        collection_id = groups["collection_id"]
        accessible_collections.extend(filter_collections(collections, owner_id if collection_id == "*" else f"{owner_id}_{collection_id}")
        
accessible_collections.extend(filter_collections(collections, user_login))`